src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -554,41 +561,35 @@ void recordWriteAfterQueryExecution() throws SQLException {
 
     public Statement getReadStatement(RouteDecisionBuilder decisionBuilder) {
         if (connectionProvider.getState().equals(MAIN)) {
-            decisionBuilder.reason(MAIN_CONNECTION_REUSE);
+            decisionBuilder.reason(MAIN_CONNECTION_REUSE);//.isWrite(true) ? ;
+            // sth was readonly, but need to happen on Main because of inconsistency
+            // I think we don't have info whether this is read  / write
             connectionProvider.getStateDecision().ifPresent(decisionBuilder::cause);
+
+
             return prepareWriteStatement(decisionBuilder);
         }
         final String sql = decisionBuilder.getSql();
-        isWriteOperation = sqlFunction.isFunctionCall(sql) || isUpdate(sql) || isDelete(sql);
+        isWriteOperation = sqlOpeartionType.isWriteOpearation(sql);
         if (isWriteOperation) {
-            decisionBuilder.reason(WRITE_OPERATION);
+            decisionBuilder.reason(WRITE_OPERATION).isWrite(true);
             return prepareWriteStatement(decisionBuilder);
         }
 
-        if (isSelectForUpdate(sql)) {
-            decisionBuilder.reason(LOCK);
+        if (sqlOpeartionType.isSelectForUpdate(sql)) {
```

Comment:
So locking is a type of a write. The `SqlQuery.writesToDatabase` should catch that. So "writing to DB" is a `Reason` to go to replica, but there are multiple types of writes. Perhaps `SqlQuery` needs to return something more than a `boolean`. Maybe it needs more explitict methods:
```java
boolean mustWriteToDb();
boolean mayWriteToDb();
String getWritingKind(); // e.g. LOCK, UPDATE, DELETE, FUNCTION
```

Reply:
What's the use case for revealing these low-level details in the API.

Reply:
I'm talking about internals. We keep hearing that "LOCK" is a type of write. Here's a way to model that explicitly.

Reply:
Make sense.

Reply:
Actually `LOCK` isn't type o write, but it's treat almost as write as it run for on main. acutall. `LOCK` == `SELECT FOR UPDATE`
I added isWrite information to Rason on compile level , this way this model say that explicitly. Thanks for this I could remove `isWriteOpearatiion` internal state, before doing this I've improved code coverage for `SELECT FOR UPDATE`.

Reply:
If it's not a write, then why is it run on main?

Reply:
I guess this is related with the impact on transaction
> FOR UPDATE causes the rows retrieved by the SELECT statement to be locked as though for update. This prevents them from being modified or deleted by other transactions until the current transaction ends.
[docs](https://www.postgresql.org/docs/9.0/sql-select.html#SQL-FOR-UPDATE-SHARE)

@wyrzyk should provide full answer on this.

Reply:
Are other connections affected by the lock? If it's not a write operation, why are they affected?

Reply:
If we send `SELECT FOR UPDATE` to replica - can it succeed?

Reply:
Do we treat connections/calls withing the same connection to  Database independently?
I think not, we had this hidden contract that read after write is treated RUN ON MAIN.
I think with `SELECT FOR UPDATE` the idea was the same.

Reply:
I'll resove conversation and move to: https://github.com/atlassian-labs/db-replica/pull/55#discussion_r590029250